gsrs-core-entities/src/main/java/ix/ginas/importers/ExcelSpreadsheetReader.java

```diff
@@ -0,0 +1,218 @@
+import java.util.stream.Stream;
+
+@Slf4j
+public class ExcelSpreadsheetReader {
+
+    private final InputStream in;
+
+    public ExcelSpreadsheetReader(InputStream inputStream) {
+        this.in = inputStream;
+        try {
+            workbook = new XSSFWorkbook(inputStream);
+        } catch (IOException e) {
+            log.error("Error initializing ExcelSpreadsheetReader", e);
+            throw new RuntimeException(e);
+        }
+    }
+
+    private final Workbook workbook;
+
+
+    public void close() throws IOException {
+        IOUtil.closeQuietly(in);
+
+        workbook.close();
+        //deletes tmp files
+        if (workbook instanceof SXSSFWorkbook) {
+            ((SXSSFWorkbook) workbook).dispose();
+        }
+    }
+
+    private static class RowWrapper implements Spreadsheet.SpreadsheetRow {
+        private final org.apache.poi.ss.usermodel.Row row;
+
+        public RowWrapper(org.apache.poi.ss.usermodel.Row r) {
+            this.row = r;
+        }
+
+
+        @Override
+        public SpreadsheetCell getCell(int j) {
+            Cell cell = row.getCell(j);
+            if (cell == null) {
+                cell = row.createCell(j);
+            }
+            return new ExcelSpreadsheetReader.CellWrapper(cell);
+        }
+    }
+
+    private static class CellWrapper implements SpreadsheetCell {
+        private final Cell cell;
+
+        public CellWrapper(Cell cell) {
+            this.cell = cell;
+        }
+
+        @Override
+        public void writeInteger(int i) {
+            cell.setCellValue(i);
+        }
+
+        @Override
+        public void writeDate(Date date) {
+            cell.setCellValue(date);
+        }
+
+        @Override
+        public void writeString(String s) {
+            cell.setCellValue(s);
+        }
+    }
+
+
+    private List<String> getValuesFromRow(Row row) {
+        List<String> fields = new ArrayList<>();
+        if (row == null) {
+            log.warn("Excel row is null!");
+            return fields;
+        }
+        row.cellIterator().forEachRemaining(c -> {
+            String fieldName = getCellValueAsString(c);
+            if (fieldName != null && fieldName.trim().length() > 0) {
+                fields.add(fieldName);
+            }
+        });
+        return fields;
+    }
+
+    public List<String> getFileFields(String sheetName, int row) {
+        Sheet fieldNameSheet = workbook.getSheet(sheetName);
+        return getFileFields(fieldNameSheet, row);
+    }
+
+    public List<String> getFileFields(Sheet fieldNameSheet, int row) {
+        Row fieldRow = fieldNameSheet.getRow(row);
+        return getValuesFromRow(fieldRow);
+    }
+
+    public List<String> getFileFields(int sheet, int row) {
+        Sheet fieldNameSheet = workbook.getSheetAt(sheet);
+        return getFileFields(fieldNameSheet, row);
+    }
+
+    public Stream<DefaultPropertyBasedRecordContext> readSheet(String sheetName, List<String> inputFields, int rowWithFields) {
```

Comment:
This method will technically work, but works by reading all rows into memory and then returning the list as a stream. This can be done asynchronously and on-demand, and should be refactored at some point. Need to add a TODO.